atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/compact/CompactPriorityCalculator.java

```diff
@@ -0,0 +1,114 @@
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.atlasdb.schema.generated.CompactMetadataTable;
+import com.palantir.atlasdb.schema.generated.CompactTableFactory;
+import com.palantir.atlasdb.schema.generated.SweepPriorityTable;
+import com.palantir.atlasdb.schema.generated.SweepTableFactory;
+import com.palantir.atlasdb.transaction.api.Transaction;
+import com.palantir.atlasdb.transaction.api.TransactionManager;
+import com.palantir.logsafe.Arg;
+
+final class CompactPriorityCalculator {
+    private static final Logger log = LoggerFactory.getLogger(CompactPriorityCalculator.class);
+
+    private final TransactionManager transactionManager;
+
+    CompactPriorityCalculator(TransactionManager transactionManager) {
+        this.transactionManager = transactionManager;
+    }
+
+    Optional<String> selectTableToCompact() {
+        return transactionManager.runTaskReadOnly(this::selectTableToCompactInternal);
+    }
+
+    private Optional<String> selectTableToCompactInternal(Transaction tx) {
+        Map<String, Long> tableToLastTimeSwept = new HashMap<>();
+        SweepPriorityTable sweepPriorityTable = SweepTableFactory.of().getSweepPriorityTable(tx);
+        sweepPriorityTable.getAllRowsUnordered(SweepPriorityTable.getColumnSelection(
+                SweepPriorityTable.SweepPriorityNamedColumn.LAST_SWEEP_TIME))
+                .forEach(row -> {
+                    Long lastSweepTime = row.getLastSweepTime();
+                    String tableName = row.getRowName().getFullTableName();
+                    tableToLastTimeSwept.put(tableName, lastSweepTime);
+                });
+
+        Map<String, Long> tableToLastTimeCompacted = new HashMap<>();
+        CompactMetadataTable compactMetadataTable = CompactTableFactory.of().getCompactMetadataTable(tx);
+        compactMetadataTable.getAllRowsUnordered(SweepPriorityTable.getColumnSelection(
+                SweepPriorityTable.SweepPriorityNamedColumn.LAST_SWEEP_TIME))
+                .forEach(row -> {
+                    Long lastCompactTime = row.getLastCompactTime();
+                    String tableName = row.getRowName().getFullTableName();
+                    tableToLastTimeCompacted.put(tableName, lastCompactTime);
+                });
+
+        List<String> uncompactedTables = tableToLastTimeSwept.keySet().stream()
+                .filter(table -> !tableToLastTimeCompacted.keySet().contains(table))
+                .collect(Collectors.toList());
+
+        if (uncompactedTables.size() > 0) {
+            log.info("There are some tables which have been swept, but not compacted. Choosing one at random.");
+            int randomTableIndex = ThreadLocalRandom.current().nextInt(uncompactedTables.size());
+            return Optional.of(uncompactedTables.get(randomTableIndex));
```

Comment:
OK. Maybe would be nice to log which one was chosen

Reply:
That's already done by the BackgroundCompactor, but I guess there's no harm in doing it here.